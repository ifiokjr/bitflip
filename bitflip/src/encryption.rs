use aes_gcm_siv::aead::Aead;
use aes_gcm_siv::aead::KeyInit;
use aes_gcm_siv::aead::OsRng;
use aes_gcm_siv::Aes256GcmSiv;
use aes_gcm_siv::Nonce;
use anyhow::Context;
use argon2::password_hash::PasswordHash;
use argon2::password_hash::PasswordHasher;
use argon2::password_hash::SaltString;
use argon2::Argon2;
use rand::RngCore;
use solana_sdk::signature::Keypair;

use crate::AppResult;

pub fn derive_key(
	encryption_secret: &[u8],
	salt: &SaltString,
	key_index: u32,
) -> AppResult<Vec<u8>> {
	let argon2 = Argon2::default();
	let encryption_secret_with_index =
		[encryption_secret, key_index.to_string().as_bytes()].concat();
	let password_hash = argon2.hash_password(&encryption_secret_with_index, salt)?;
	let key = PasswordHash::new(&password_hash.to_string())?
		.hash
		.ok_or(anyhow::anyhow!("could not unwrap hash"))?;
	let result = key.as_bytes().to_vec();

	Ok(result)
}

pub(crate) fn encrypt_keypair(
	keypair: &Keypair,
	encryption_key: &[u8],
	nonce: &Nonce,
) -> AppResult<Vec<u8>> {
	let cipher = Aes256GcmSiv::new(encryption_key.into());
	let serialized_keypair = keypair.to_bytes();

	let ciphertext = cipher
		.encrypt(nonce, serialized_keypair.as_ref())
		.context("encryption failure!")?;

	Ok(ciphertext)
}

pub(crate) fn decrypt_keypair(
	ciphertext: &[u8],
	decryption_key: &[u8],
	nonce: &Nonce,
) -> AppResult<Keypair> {
	let cipher = Aes256GcmSiv::new(decryption_key.into());
	let decrypted_bytes = cipher
		.decrypt(nonce, ciphertext.as_ref())
		.context("decryption failure!")?;
	let keypair = Keypair::from_bytes(&decrypted_bytes).context("keypair generation failed")?;

	Ok(keypair)
}

pub(crate) fn generate_nonce() -> Nonce {
	let mut nonce_bytes = [0u8; 12];
	OsRng.fill_bytes(&mut nonce_bytes);

	*Nonce::from_slice(&nonce_bytes)
}

pub(crate) fn generate_salt() -> SaltString {
	SaltString::generate(&mut OsRng)
}
