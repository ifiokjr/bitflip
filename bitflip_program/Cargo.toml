[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
