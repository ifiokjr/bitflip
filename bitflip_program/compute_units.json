{
	"instructions": {
		"config_initialize": {
			"compute_units": 17569,
			"rounded_compute_units": 20000,
			"description": "Initialize the config"
		},
		"config_update_authority": {
			"compute_units": 2326,
			"rounded_compute_units": 10000,
			"description": "Update the authority of the config"
		},
		"flip_bit_test": {
			"compute_units": 64832,
			"rounded_compute_units": 70000,
			"description": "Flip a single bit from `0` to `1`"
		},
		"game_initialize": {
			"compute_units": 12979,
			"rounded_compute_units": 20000,
			"description": "Initialize the game"
		},
		"game_refresh_signer": {
			"compute_units": 2232,
			"rounded_compute_units": 10000,
			"description": "Refresh the signer"
		},
		"token_group_initialize": {
			"compute_units": 47429,
			"rounded_compute_units": 50000,
			"description": "Initialize the token group"
		},
		"token_initialize:bit": {
			"compute_units": 63184,
			"rounded_compute_units": 70000,
			"description": "Initialize group token: Bit"
		},
		"token_initialize:gibibit": {
			"compute_units": 59746,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: GibiBit"
		},
		"token_initialize:kibibit": {
			"compute_units": 59745,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: KibiBit"
		},
		"token_initialize:mebibit": {
			"compute_units": 59742,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: MebiBit"
		}
	},
	"version": "0.0.0"
}
