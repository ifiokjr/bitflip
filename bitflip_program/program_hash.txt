ebc22d77bc8ddd23f326598673e9fe7aa99fcdce9572fb8559a8d9782ade4ab8
