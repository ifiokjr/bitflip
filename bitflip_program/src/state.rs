use solana_program::msg;
use steel::*;

use crate::BITFLIP_SECTION_LENGTH;
use crate::BITFLIP_SECTION_TOTAL_BITS;
use crate::SESSION_DURATION;

#[repr(u8)]
#[derive(Clone, Copy, Debug, Eq, PartialEq, IntoPrimitive, TryFromPrimitive)]
pub enum BitflipAccount {
	ConfigState = 0,
	GameState = 1,
	SectionState = 2,
}

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable)]
#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
#[cfg_attr(feature = "client", derive(typed_builder::TypedBuilder))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct ConfigState {
	/// The authority which can update this config.
	#[cfg_attr(
		feature = "serde",
		serde(with = "::serde_with::As::<serde_with::DisplayFromStr>")
	)]
	pub authority: Pubkey,
	/// Store the bump to save compute units.
	pub bump: u8,
	/// The treasury account bump where fees are sent and where the minted
	/// tokens are transferred.
	pub treasury_bump: u8,
	/// The mint account bump.
	pub mint_bit_bump: u8,
	/// There will be a maximum of 256 games.
	pub game_index: u8,
}

impl ConfigState {
	pub const fn space() -> usize {
		8 + std::mem::size_of::<Self>()
	}

	pub fn new(authority: Pubkey, bump: u8, treasury_bump: u8, mint_bit_bump: u8) -> ConfigState {
		ConfigState {
			authority,
			bump,
			treasury_bump,
			mint_bit_bump,
			game_index: 0,
		}
	}
}

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "client", derive(typed_builder::TypedBuilder))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct GameState {
	/// This is a refresh signer created and maintained by the backend. It needs
	/// to be provided to update the access signer. It will need to be updated
	/// after every game.
	#[cfg_attr(
		feature = "serde",
		serde(with = "::serde_with::As::<serde_with::DisplayFromStr>")
	)]
	#[cfg_attr(feature = "client", builder(default))]
	pub refresh_signer: Pubkey,
	/// This is an access signer created and maintained by the backend. Which is
	/// allowed to sign certain transactions and expires daily.
	#[cfg_attr(
		feature = "serde",
		serde(with = "::serde_with::As::<serde_with::DisplayFromStr>")
	)]
	#[cfg_attr(feature = "client", builder(default))]
	pub access_signer: Pubkey,
	/// The timestamp that the access expiry will end.
	#[cfg_attr(feature = "client", builder(default))]
	pub access_expiry: i64,
	/// The start time. If 0 then it hasn't started yet. Using an `Option` here
	/// would waste an extra byte.
	#[cfg_attr(feature = "client", builder(default))]
	pub start_time: i64,
	/// The index of this currently played game.
	pub game_index: u8,
	/// The most recent section which was unlocked. This will be updated every
	/// time a new section is initialized.
	pub section_index: u8,
	/// The bump for this account.
	pub bump: u8,
	/// The bump for the nonce account.
	pub nonce_bump: u8,
	/// Padding to make the size of the struct a multiple of 8.
	#[cfg_attr(feature = "client", builder(default))]
	pub _padding: [u8; 4],
}

impl GameState {
	pub fn new(
		access_signer: Pubkey,
		refresh_signer: Pubkey,
		index: u8,
		bump: u8,
		nonce_bump: u8,
	) -> Self {
		Self {
			refresh_signer,
			access_signer,
			access_expiry: 0,
			start_time: 0,
			section_index: 0,
			game_index: index,
			bump,
			nonce_bump,
			_padding: [0; 4],
		}
	}

	pub fn end_time(&self) -> i64 {
		self.start_time.saturating_add(SESSION_DURATION)
	}

	pub fn started(&self) -> bool {
		msg!("start_time: {}", self.start_time);
		self.start_time > 0
	}

	pub fn ended(&self, current_time: i64) -> bool {
		msg!(
			"current_time: {}, end_time: {}",
			current_time,
			self.end_time()
		);
		self.started() && current_time > self.end_time()
	}

	pub fn running(&self, current_time: i64) -> bool {
		self.started() && !self.ended(current_time)
	}
}

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
pub struct SectionState {
	/// The state of the bits that are represented as flippable bits on the
	/// frontend.
	#[cfg_attr(feature = "serde", serde(with = "serde_big_array::BigArray"))]
	pub data: [u16; BITFLIP_SECTION_LENGTH],
	/// The owner of this section.
	#[cfg_attr(
		feature = "serde",
		serde(with = "::serde_with::As::<serde_with::DisplayFromStr>")
	)]
	pub owner: Pubkey,
	/// The number of bit flips that have occurred.
	pub flips: u32,
	/// The number of bits that are on.
	pub on: u32,
	/// The number of bits that are off.
	pub off: u32,
	/// The index for this section state.
	pub index: u8,
	/// The bump for this section state.
	pub bump: u8,
	/// Padding to make the size of the struct a multiple of 8.
	pub _padding: [u8; 2],
}

impl SectionState {
	/// Create a new section state in the client. Useful for testing.
	pub fn new(owner: Pubkey, index: u8, bump: u8) -> Self {
		Self {
			data: [0; BITFLIP_SECTION_LENGTH],
			owner,
			flips: 0,
			on: 0,
			off: BITFLIP_SECTION_TOTAL_BITS,
			bump,
			index,
			_padding: [0; 2],
		}
	}

	/// Initialize the section state without touching the data to prevent using
	/// compute units.
	pub fn init(&mut self, owner: Pubkey, index: u8, bump: u8) {
		self.owner = owner;
		self.index = index;
		self.bump = bump;
		self.on = BITFLIP_SECTION_TOTAL_BITS;
		self.off = 0;
		self.flips = 0;
	}

	pub fn flip_on(&mut self, changed_bits: u32) -> ProgramResult {
		self.on = self
			.on
			.checked_add(changed_bits)
			.ok_or(ProgramError::ArithmeticOverflow)?;
		self.off = self
			.off
			.checked_sub(changed_bits)
			.ok_or(ProgramError::ArithmeticOverflow)?;
		self.flips = self
			.flips
			.checked_add(changed_bits)
			.ok_or(ProgramError::ArithmeticOverflow)?;

		Ok(())
	}

	pub fn flip_off(&mut self, changed_bits: u32) -> ProgramResult {
		self.off = self
			.off
			.checked_add(changed_bits)
			.ok_or(ProgramError::ArithmeticOverflow)?;
		self.on = self
			.on
			.checked_sub(changed_bits)
			.ok_or(ProgramError::ArithmeticOverflow)?;
		self.flips = self
			.flips
			.checked_add(changed_bits)
			.ok_or(ProgramError::ArithmeticOverflow)?;

		Ok(())
	}
}

account!(BitflipAccount, ConfigState);
account!(BitflipAccount, GameState);
account!(BitflipAccount, SectionState);
