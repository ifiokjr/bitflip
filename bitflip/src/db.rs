use std::str::FromStr;
#[cfg(feature = "ssr")]
use std::sync::Arc;

#[cfg(feature = "ssr")]
use anyhow::Context;
use bitflip_program::GameStatus;
use bitflip_program::SectionData;
use bitflip_program::BITFLIP_SECTION_LENGTH;
use chrono::DateTime;
use chrono::Utc;
use num_enum::IntoPrimitive;
use num_enum::TryFromPrimitive;
use rand::rngs::OsRng;
use serde::Deserialize;
use serde::Serialize;
use solana_sdk::pubkey::Pubkey;
use solana_sdk::signature::Keypair;
use uuid::Uuid;
#[cfg(feature = "ssr")]
use welds::connections::sqlite::connect;
#[cfg(feature = "ssr")]
use welds::connections::sqlite::SqliteClient;
#[cfg(feature = "ssr")]
use welds::state::DbState;
#[cfg(feature = "ssr")]
use welds::WeldsModel;

use crate::encryption::decrypt_keypair;
use crate::encryption::derive_key;
use crate::AppResult;
use crate::AppStateConfig;

#[cfg(feature = "ssr")]
#[derive(derive_more::Debug, Clone, derive_more::From, derive_more::Into, derive_more::Deref)]
pub struct Db(#[debug(skip)] Arc<SqliteClient>);

impl Db {
	#[cfg(feature = "ssr")]
	pub async fn try_new(connection_string: &str) -> AppResult<Self> {
		let connection = connect(connection_string)
			.await
			.context("could not connect to the database")?;

		Ok(Arc::new(connection).into())
	}

	#[cfg(feature = "ssr")]
	pub fn as_sqlx_pool(&self) -> &sqlx::SqlitePool {
		self.0.as_sqlx_pool()
	}

	#[cfg(feature = "ssr")]
	pub fn as_client(&self) -> &SqliteClient {
		self.0.as_ref()
	}
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ssr", derive(WeldsModel))]
#[cfg_attr(feature = "ssr", welds(table = "games"))]
#[cfg_attr(feature = "ssr", welds(HasMany(section, Section, "game_index")))]
pub struct Game {
	#[cfg_attr(feature = "ssr", welds(primary_key))]
	id: Uuid,
	game_index: u8,
	start_time: DateTime<Utc>,
	duration: i64,
	min_lamports: i64,
	base_lamports: i64,
	max_lamports: i64,
	status: u8,
	section_index: u8,
	game_signer: String,
	temp_signer: String,
	created_at: DateTime<Utc>,
	updated_at: DateTime<Utc>,
}

impl Game {
	pub fn set_id(&mut self) -> &mut Self {
		self.id = Uuid::now_v7();
		self
	}

	pub fn set_game_index(&mut self, game_index: u8) -> &mut Self {
		self.game_index = game_index;
		self
	}

	pub fn set_start_time(&mut self, start_time: DateTime<Utc>) -> &mut Self {
		self.start_time = start_time;
		self
	}

	pub fn set_duration(&mut self, duration: i64) -> &mut Self {
		self.duration = duration;
		self
	}

	pub fn set_min_lamports(&mut self, lamports: u64) -> &mut Self {
		self.min_lamports = lamports as i64;
		self
	}

	pub fn set_base_lamports(&mut self, lamports: u64) -> &mut Self {
		self.base_lamports = lamports as i64;
		self
	}

	pub fn set_max_lamports(&mut self, lamports: u64) -> &mut Self {
		self.max_lamports = lamports as i64;
		self
	}

	pub fn set_status(&mut self, status: GameStatus) -> &mut Self {
		self.status = status.into();
		self
	}

	pub fn set_section_index(&mut self, section_index: u8) -> &mut Self {
		self.section_index = section_index;
		self
	}

	/// Store the permanent `game_signer`
	pub fn set_game_signer(&mut self, pubkey: &Pubkey) -> &mut Self {
		self.game_signer = pubkey.to_string();
		self
	}

	/// Set the temporary signer pubkey for the game.
	pub fn set_temp_signer(&mut self, pubkey: &Pubkey) -> &mut Self {
		self.temp_signer = pubkey.to_string();
		self
	}

	pub fn id(&self) -> Uuid {
		self.id
	}

	pub fn game_index(&self) -> u8 {
		self.game_index
	}

	pub fn start_time(&self) -> DateTime<Utc> {
		self.start_time
	}

	pub fn duration(&self) -> i64 {
		self.duration
	}

	pub fn min_lamports(&self) -> u64 {
		self.min_lamports as u64
	}

	pub fn base_lamports(&self) -> u64 {
		self.base_lamports as u64
	}

	pub fn max_lamports(&self) -> u64 {
		self.max_lamports as u64
	}

	pub fn status(&self) -> GameStatus {
		GameStatus::try_from(self.status).unwrap_or(GameStatus::Pending)
	}

	pub fn section_index(&self) -> u8 {
		self.section_index
	}

	pub fn game_signer(&self) -> Pubkey {
		Pubkey::from_str_const(&self.game_signer)
	}

	pub fn temp_signer(&self) -> Pubkey {
		Pubkey::from_str_const(&self.temp_signer)
	}

	pub fn created_at(&self) -> DateTime<Utc> {
		self.created_at
	}

	pub fn updated_at(&self) -> DateTime<Utc> {
		self.updated_at
	}
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ssr", derive(WeldsModel))]
#[cfg_attr(feature = "ssr", welds(table = "sections"))]
#[cfg_attr(
	feature = "ssr",
	welds(HasMany(section_event, SectionEvent, "section_id"))
)]
#[cfg_attr(feature = "ssr", welds(BelongsTo(game, Game, "game_index")))]
pub struct Section {
	#[cfg_attr(feature = "ssr", welds(primary_key))]
	id: Uuid,
	game_index: u8,
	section_index: u8,
	data: Vec<u8>,
	reward_tokens: i64,
	lamports: i64,
	owner: String,
	flips: u32,
	created_at: DateTime<Utc>,
	updated_at: DateTime<Utc>,
}

impl Section {
	pub fn set_id(&mut self) -> &mut Self {
		self.id = Uuid::now_v7();
		self
	}

	pub fn set_game_index(&mut self, game_index: u8) -> &mut Self {
		self.game_index = game_index;
		self
	}

	pub fn set_section_index(&mut self, section_index: u8) -> &mut Self {
		self.section_index = section_index;
		self
	}

	pub fn set_data(&mut self, data: SectionData) -> &mut Self {
		self.data = {
			let v: &SectionData = &data;
			v.iter().flat_map(|x| x.to_le_bytes()).collect::<Vec<u8>>()
		};
		self
	}

	pub fn set_reward_tokens(&mut self, reward_tokens: u64) -> &mut Self {
		self.reward_tokens = reward_tokens as i64;
		self
	}

	pub fn set_lamports(&mut self, lamports: u64) -> &mut Self {
		self.lamports = lamports as i64;
		self
	}

	pub fn set_owner(&mut self, owner: Pubkey) -> &mut Self {
		self.owner = owner.to_string();
		self
	}

	pub fn set_flips(&mut self, flips: u32) -> &mut Self {
		self.flips = flips;
		self
	}

	pub fn id(&self) -> Uuid {
		self.id
	}

	pub fn game_index(&self) -> u8 {
		self.game_index
	}

	pub fn section_index(&self) -> u8 {
		self.section_index
	}

	pub fn data(&self) -> SectionData {
		let bytes: &[u8] = &self.data;
		assert_eq!(bytes.len(), 512, "Input bytes must be exactly 512 bytes");
		let mut result = [0u16; BITFLIP_SECTION_LENGTH];

		for (ii, chunk) in bytes.chunks_exact(2).enumerate() {
			result[ii] = u16::from_le_bytes([chunk[0], chunk[1]]);
		}

		result
	}

	pub fn reward_tokens(&self) -> u64 {
		self.reward_tokens as u64
	}

	pub fn lamports(&self) -> u64 {
		self.lamports as u64
	}

	pub fn owner(&self) -> Pubkey {
		Pubkey::from_str(&self.owner).unwrap()
	}

	pub fn flips(&self) -> u32 {
		self.flips
	}

	pub fn created_at(&self) -> DateTime<Utc> {
		self.created_at
	}

	pub fn updated_at(&self) -> DateTime<Utc> {
		self.updated_at
	}
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ssr", derive(WeldsModel))]
#[cfg_attr(feature = "ssr", welds(table = "section_events"))]
#[cfg_attr(feature = "ssr", welds(BelongsTo(player, Player, "player_pubkey")))]
#[cfg_attr(feature = "ssr", welds(BelongsTo(section, Section, "section_id")))]
pub struct SectionEvent {
	#[cfg_attr(feature = "ssr", welds(primary_key))]
	id: Uuid,
	section_id: Uuid,
	player_pubkey: String,
	event_type: u8,
	data: Vec<u8>,
	created_at: DateTime<Utc>,
}

#[repr(u8)]
#[derive(Default, Clone, Copy, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
#[non_exhaustive]
pub enum SectionEventType {
	#[default]
	None,
	Color,
	Minesweeper,
}

impl SectionEvent {
	pub fn set_id(&mut self) -> &mut Self {
		self.id = Uuid::now_v7();
		self
	}

	pub fn set_section_id(&mut self, section_id: Uuid) -> &mut Self {
		self.section_id = section_id;
		self
	}

	pub fn set_player_pubkey(&mut self, player_pubkey: Pubkey) -> &mut Self {
		self.player_pubkey = player_pubkey.to_string();
		self
	}

	pub fn set_event_type(&mut self, event_type: SectionEventType) -> &mut Self {
		self.event_type = event_type.into();
		self
	}

	pub fn set_data(&mut self, data: Vec<u8>) -> &mut Self {
		self.data = data;
		self
	}

	pub fn id(&self) -> Uuid {
		self.id
	}

	pub fn section_id(&self) -> Uuid {
		self.section_id
	}

	pub fn player_pubkey(&self) -> Pubkey {
		Pubkey::from_str(&self.player_pubkey).unwrap()
	}

	pub fn event_type(&self) -> SectionEventType {
		SectionEventType::try_from(self.event_type).unwrap_or(SectionEventType::None)
	}

	pub fn data(&self) -> &[u8] {
		&self.data
	}

	pub fn created_at(&self) -> DateTime<Utc> {
		self.created_at
	}
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ssr", derive(WeldsModel))]
#[cfg_attr(feature = "ssr", welds(table = "players"))]
#[cfg_attr(
	feature = "ssr",
	welds(HasMany(section_event, SectionEvent, "player_pubkey"))
)]
pub struct Player {
	#[welds(primary_key)]
	pubkey: String,
	signer: String,
	updated_at: DateTime<Utc>,
	created_at: DateTime<Utc>,
}

impl Player {
	pub fn set_pubkey(&mut self, pubkey: Pubkey) -> &mut Self {
		self.pubkey = pubkey.to_string();
		self
	}

	pub fn set_signer(&mut self, signer: &Pubkey) -> &mut Self {
		self.signer = signer.to_string();
		self
	}

	pub fn pubkey(&self) -> Pubkey {
		Pubkey::from_str(&self.pubkey).unwrap()
	}

	pub fn signer(&self) -> Pubkey {
		Pubkey::from_str_const(&self.signer)
	}

	pub fn created_at(&self) -> DateTime<Utc> {
		self.created_at
	}

	pub fn updated_at(&self) -> DateTime<Utc> {
		self.updated_at
	}
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "ssr", derive(WeldsModel))]
#[cfg_attr(feature = "ssr", welds(table = "encrypted_keypairs"))]
pub struct EncryptedKeypair {
	#[welds(primary_key)]
	pubkey: String,
	encrypted_keypair: Vec<u8>,
	key_index: u32,
	nonce: Vec<u8>,
	salt: String,
	status: u8,
	created_at: DateTime<Utc>,
	updated_at: DateTime<Utc>,
}

impl EncryptedKeypair {
	#[cfg(feature = "ssr")]
	pub async fn find_by_pubkey(
		client: &SqliteClient,
		pubkey: &Pubkey,
	) -> AppResult<Option<DbState<Self>>> {
		let result = Self::find_by_id(client, pubkey.to_string()).await?;

		Ok(result)
	}

	#[cfg(feature = "ssr")]
	pub fn generate(&mut self, config: &AppStateConfig) -> AppResult<Keypair> {
		use rand::Rng;
		use solana_sdk::signer::Signer;

		use crate::encryption::encrypt_keypair;
		use crate::encryption::generate_nonce;
		use crate::encryption::generate_salt;

		let key_index = OsRng
			.gen_range(0..config.max_active_keys)
			.saturating_add(config.key_multiplier.saturating_mul(config.max_active_keys));
		let keypair = Keypair::new();
		let nonce = generate_nonce();
		let salt = generate_salt();
		let encryption_key = derive_key(config.encryption_secret.as_bytes(), &salt, key_index)?;
		let encrypted_keypair = encrypt_keypair(&keypair, &encryption_key, &nonce)?;

		self.pubkey = keypair.pubkey().to_string();
		self.encrypted_keypair = encrypted_keypair;
		self.key_index = key_index;
		self.nonce = nonce.to_vec();
		self.salt = salt.to_string();
		self.status = EncryptedKeypairStatus::Active.into();

		Ok(keypair)
	}

	pub fn pubkey(&self) -> Pubkey {
		Pubkey::from_str(&self.pubkey).unwrap()
	}

	pub fn encrypted_keypair(&self) -> &[u8] {
		&self.encrypted_keypair
	}

	#[cfg(feature = "ssr")]
	pub fn keypair(&self, config: &AppStateConfig) -> AppResult<Keypair> {
		use aes_gcm_siv::Nonce;
		use argon2::password_hash::SaltString;

		let nonce = Nonce::from_slice(self.nonce());
		let salt = SaltString::from_b64(self.salt())?;
		let key_index = self.key_index;
		let encrypted_keypair = &self.encrypted_keypair;
		let decryption_key = derive_key(config.encryption_secret.as_bytes(), &salt, key_index)?;
		let result = decrypt_keypair(encrypted_keypair, &decryption_key, nonce)?;

		Ok(result)
	}

	pub fn nonce(&self) -> &[u8] {
		&self.nonce
	}

	pub fn salt(&self) -> &str {
		&self.salt
	}

	pub fn status(&self) -> EncryptedKeypairStatus {
		EncryptedKeypairStatus::try_from(self.status).unwrap_or(EncryptedKeypairStatus::Active)
	}

	pub fn created_at(&self) -> DateTime<Utc> {
		self.created_at
	}

	pub fn updated_at(&self) -> DateTime<Utc> {
		self.updated_at
	}
}

#[repr(u8)]
#[derive(Default, Clone, Copy, Debug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
#[non_exhaustive]
pub enum EncryptedKeypairStatus {
	/// The keypair is active and can be used
	#[default]
	Active,
	/// The keypair has been revoked and emptied.
	Revoked,
	/// The keypair has been compromised and should not be used
	Compromised,
}

#[cfg(all(feature = "ssr", test))]
mod tests {
	use solana_sdk::signer::Signer;
	use test_log::test;

	use super::*;
	use crate::AppEnvironment;

	async fn setup_db() -> AppResult<Db> {
		let db = Db::try_new("sqlite::memory:").await?;
		sqlx::migrate!("../migrations")
			.run(db.as_sqlx_pool())
			.await?;

		Ok(db)
	}

	#[test(tokio::test)]
	async fn games_table_snapshot() -> AppResult<()> {
		let db = setup_db().await?;
		let diff = welds::check::schema::<Game>(db.as_ref()).await?;
		insta::assert_debug_snapshot!(diff);

		Ok(())
	}

	#[test(tokio::test)]
	async fn sections_table_snapshot() -> AppResult<()> {
		let db = setup_db().await?;
		let diff = welds::check::schema::<Section>(db.as_ref()).await?;
		insta::assert_debug_snapshot!(diff);

		Ok(())
	}

	#[test(tokio::test)]
	async fn section_events_table_snapshot() -> AppResult<()> {
		let db = setup_db().await?;
		let diff = welds::check::schema::<SectionEvent>(db.as_ref()).await?;
		insta::assert_debug_snapshot!(diff);

		Ok(())
	}

	#[test(tokio::test)]
	async fn players_table_snapshot() -> AppResult<()> {
		let db = setup_db().await?;
		let diff = welds::check::schema::<Player>(db.as_ref()).await?;
		insta::assert_debug_snapshot!(diff);

		Ok(())
	}

	#[test(tokio::test)]
	async fn encrypted_keypairs_table_snapshot() -> AppResult<()> {
		let db = setup_db().await?;
		let diff = welds::check::schema::<EncryptedKeypair>(db.as_ref()).await?;
		insta::assert_debug_snapshot!(diff);

		Ok(())
	}

	#[test(tokio::test)]
	async fn encrypted_keypair_generate() -> AppResult<()> {
		let db = setup_db().await?;
		let client = db.as_client();
		let config = AppStateConfig {
			encryption_secret: "test encryption secret".to_string(),
			key_multiplier: 0,
			max_active_keys: 512,
			environment: AppEnvironment::Local,
			database_url: "sqlite::memory:".parse()?,
			website_url: "http://localhost:3000".parse()?,
		};
		let mut encrypted_keypair = EncryptedKeypair::new();
		let keypair = encrypted_keypair.generate(&config)?;
		encrypted_keypair.save(client).await?;

		let result = EncryptedKeypair::find_by_id(client, keypair.pubkey().to_string())
			.await?
			.unwrap();

		assert_eq!(result.keypair(&config)?, keypair);

		Ok(())
	}
}
