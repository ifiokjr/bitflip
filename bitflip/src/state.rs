use std::sync::Arc;

use axum::extract::FromRef;
use leptos::config::LeptosOptions;
use typed_builder::TypedBuilder;
use url::Url;

use crate::db::Db;
use crate::AppEnvironment;

#[derive(FromRef, Debug, Clone, TypedBuilder)]
pub struct AppState {
	pub leptos: LeptosOptions,
	pub config: Arc<AppStateConfig>,
	pub db: Db,
}

#[derive(Debug, Clone)]
pub struct AppStateConfig {
	/// The encryption secret used to derive the encryption keys for keypairs.
	pub encryption_secret: String,
	/// The multiplier for the key index which is used to set the band that the
	/// `key_index` can fall within.
	pub key_multiplier: u32,
	/// The number of keys that can be active at a time.
	pub max_active_keys: u32,
	/// The environment that the app is running in.
	pub environment: AppEnvironment,
	/// The database URL.
	pub database_url: Url,
	/// The website URL.
	pub website_url: Url,
}
