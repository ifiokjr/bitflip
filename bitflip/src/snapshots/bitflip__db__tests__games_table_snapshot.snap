---
source: bitflip/src/db.rs
expression: diff
snapshot_kind: text
---
[
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: Medium,
        kind: Changed(
            Diff {
                column: "id",
                db_type: "TEXT",
                db_nullable: true,
                welds_type: "Uuid",
                welds_nullable: false,
                type_changed: false,
            },
        ),
    },
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: High,
        kind: Changed(
            Diff {
                column: "game_index",
                db_type: "INTEGER",
                db_nullable: false,
                welds_type: "u8",
                welds_nullable: false,
                type_changed: true,
            },
        ),
    },
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: High,
        kind: Changed(
            Diff {
                column: "start_time",
                db_type: "INTEGER",
                db_nullable: false,
                welds_type: "DateTime<Utc>",
                welds_nullable: false,
                type_changed: true,
            },
        ),
    },
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: High,
        kind: Changed(
            Diff {
                column: "status",
                db_type: "INTEGER",
                db_nullable: false,
                welds_type: "u8",
                welds_nullable: false,
                type_changed: true,
            },
        ),
    },
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: High,
        kind: Changed(
            Diff {
                column: "section_index",
                db_type: "INTEGER",
                db_nullable: false,
                welds_type: "u8",
                welds_nullable: false,
                type_changed: true,
            },
        ),
    },
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: High,
        kind: Changed(
            Diff {
                column: "created_at",
                db_type: "TEXT",
                db_nullable: false,
                welds_type: "DateTime<Utc>",
                welds_nullable: false,
                type_changed: true,
            },
        ),
    },
    Issue {
        ident: TableIdent {
            schema: None,
            name: "games",
        },
        level: High,
        kind: Changed(
            Diff {
                column: "updated_at",
                db_type: "TEXT",
                db_nullable: false,
                welds_type: "DateTime<Utc>",
                welds_nullable: false,
                type_changed: true,
            },
        ),
    },
]
